src/mpi/coll/transports/mpich/transport.h

```diff
@@ -170,17 +170,27 @@ static inline void MPIC_MPICH_free_buffers(MPIC_MPICH_sched_t * s)
         /*free the temporary memory allocated by recv_reduce call */
         if (s->requests[i].kind == MPIC_MPICH_KIND_RECV_REDUCE) {
             MPIC_MPICH_free_mem(s->requests[i].nbargs.recv_reduce.inbuf);
+            s->requests[i].nbargs.recv_reduce.inbuf = NULL;
         }
     }
     /*free temporary buffers */
     for (i = 0; i < s->nbufs; i++) {
         MPIC_MPICH_free_mem(s->buf[i]);
+        s->buf[i] = NULL;
     }
 }
 
 static inline void MPIC_MPICH_sched_destroy_fn(MPIC_MPICH_sched_t * s)
 {
     MPIC_MPICH_free_buffers(s);
+    int i;
+    for (i = 0; i < s->total; i++) {
+        /*Really deallocate memory marked for deallocation*/
+        if (s->requests[i].kind == MPIC_MPICH_KIND_FREE_MEM) {
+            MPIC_MPICH_free_mem(s->requests[i].nbargs.free_mem.ptr);
+            s->requests[i].nbargs.free_mem.ptr = NULL;
+        }
+    }
```

Comment:
Please move this for loop inside MPIC_MPICH_free_buffers. There is already a for loop there, you can put this as an else if condition in that existing for loop